staging/src/k8s.io/apimachinery/pkg/util/cache/lruexpirecache.go

```diff
@@ -67,36 +69,81 @@ type cacheEntry struct {
 func (c *LRUExpireCache) Add(key interface{}, value interface{}, ttl time.Duration) {
 	c.lock.Lock()
 	defer c.lock.Unlock()
-	c.cache.Add(key, &cacheEntry{value, c.clock.Now().Add(ttl)})
+
+	// Key already exists
+	oldElement, ok := c.entries[key]
+	if ok {
+		c.evictionList.MoveToFront(oldElement)
+		oldElement.Value.(*cacheEntry).value = value
+		oldElement.Value.(*cacheEntry).expireTime = c.clock.Now().Add(ttl)
+		return
+	}
+
+	// Make space if necessary
+	if c.evictionList.Len() >= c.maxSize {
+		toEvict := c.evictionList.Back()
+		c.evictionList.Remove(toEvict)
+		delete(c.entries, toEvict.Value.(*cacheEntry).key)
+	}
+
+	// Add new entry
+	entry := &cacheEntry{
+		key:        key,
+		value:      value,
+		expireTime: c.clock.Now().Add(ttl),
+	}
+	element := c.evictionList.PushFront(entry)
+	c.entries[key] = element
 }
 
 // Get returns the value at the specified key from the cache if it exists and is not
 // expired, or returns false.
 func (c *LRUExpireCache) Get(key interface{}) (interface{}, bool) {
 	c.lock.Lock()
 	defer c.lock.Unlock()
-	e, ok := c.cache.Get(key)
+
+	element, ok := c.entries[key]
 	if !ok {
 		return nil, false
 	}
-	if c.clock.Now().After(e.(*cacheEntry).expireTime) {
-		c.cache.Remove(key)
+
+	if c.clock.Now().After(element.Value.(*cacheEntry).expireTime) {
+		c.evictionList.Remove(element)
+		delete(c.entries, element.Value.(*cacheEntry).key)
 		return nil, false
 	}
-	return e.(*cacheEntry).value, true
+
+	c.evictionList.MoveToFront(element)
+
+	return element.Value.(*cacheEntry).value, true
 }
 
 // Remove removes the specified key from the cache if it exists
 func (c *LRUExpireCache) Remove(key interface{}) {
 	c.lock.Lock()
 	defer c.lock.Unlock()
-	c.cache.Remove(key)
+
+	element, ok := c.entries[key]
+	if !ok {
+		return
+	}
+
+	c.evictionList.Remove(element)
+	delete(c.entries, element.Value.(*cacheEntry).key)
 }
 
 // Keys returns all the keys in the cache, even if they are expired. Subsequent calls to
 // get may return not found. It returns all keys from oldest to newest.
 func (c *LRUExpireCache) Keys() []interface{} {
 	c.lock.Lock()
 	defer c.lock.Unlock()
-	return c.cache.Keys()
+
+	val := []interface{}{}
```

Comment:
pre-allocate the list with the right capacity since we know the number of items we'll return

Reply:
Done.